File: frontend/src/types/scan.ts
/** Mirrors backend/app/schemas/scan.py */

export type ScanStatus = 'QUEUED' | 'RUNNING' | 'COMPLETED' | 'FAILED'

/** A scan that has not finished — the UI polls while one of these is true. */
export const ACTIVE_SCAN_STATUSES: ScanStatus[] = ['QUEUED', 'RUNNING']

export interface Scan {
  id: number
  repository_id: number
  status: ScanStatus
  attempts: number
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  files_scanned: number
  files_skipped: number
  unparsable_files: number
  /** What this run concluded, frozen when it ran. */
  total_findings: number
  new_findings: number
  fixed_findings: number
  truncated: boolean
  error_message: string | null
  created_at: string
}

export interface ScanListResponse {
  items: Scan[]
  total: number
  limit: number
  offset: number
}

File: backend/app/schemas/scan.py
"""Scan request/response schemas.

A queued scan is returned with `202 Accepted`, not `200`: the work has been
accepted, not done. The client polls `GET /scans/{id}` until the status leaves
`QUEUED`/`RUNNING`.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.scan import ScanStatus


class ScanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    repository_id: int
    status: ScanStatus
    attempts: int
    started_at: datetime | None
    finished_at: datetime | None
    duration_ms: int | None
    files_scanned: int
    files_skipped: int
    unparsable_files: int
    # What this run concluded, frozen at the time it ran — a history row must
    # still be true after the findings have moved on.
    total_findings: int
    new_findings: int
    fixed_findings: int
    truncated: bool
    error_message: str | None
    created_at: datetime


class ScanListResponse(BaseModel):
    items: list[ScanRead]
    total: int
    limit: int
    offset: int
